habitat-lab/habitat/sims/habitat_simulator/object_state_machine.py

```diff
@@ -0,0 +1,338 @@
+        if obj_state:
+            color = mn.Color4.green()
+
+        draw_object_highlight(obj, debug_line_render, camera_transform, color)
+
+    def toggle(
+        self, obj: Union[ManagedArticulatedObject, ManagedRigidObject]
+    ) -> bool:
+        """
+        Toggles a boolean state, returning the newly set value.
+
+        :param obj: The ManagedObject instance.
+
+        :return: The new value of the state.
+        """
+
+        cur_state = get_state_of_obj(obj, self.name)
+        new_state = not cur_state
+        set_state_of_obj(obj, self.name, new_state)
+        return new_state
+
+
+class ObjectIsClean(BooleanObjectState):
+    """
+    ObjectIsClean state specifies whether an object is clean or dirty.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.name = "is_clean"
+        # TODO: set the semantic class membership list
+        self.accepted_semantic_classes = []
+
+
+class ObjectIsPoweredOn(BooleanObjectState):
+    """
+    State specifies whether an appliance object is powered on or off.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.name = "is_powered_on"
+        # TODO: set the semantic class membership list
+        self.accepted_semantic_classes = []
+
+    def default_value(self) -> Any:
+        """
+        Default value for power is off.
+        """
+
+        return False
+
+
+class ObjectSateMachine:
+    """
+    Defines the logic for managing multiple states across all objects in the scene.
+    """
+
+    def __init__(self, active_states: List[ObjectStateSpec] = None) -> None:
+        # a list of ObjectStateSpec singleton instances which are active in the current scene
+        self.active_states = active_states if active_states is not None else []
+        # map tracked objects to their set of state properies
+        self.objects_with_states: Dict[str, List[ObjectStateSpec]] = {}
+
+    def initialize_object_state_map(self, sim: habitat_sim.Simulator) -> None:
+        """
+        Reset the objects_with_states dict and re-initializes it by parsing all objects from the scene and checking is_affordance_of_obj for all active ObjectStateSpecs.
+
+        :param sim: The Simulator instance.
+        """
+
+        self.objects_with_states = {}
+        all_objects = sutils.get_all_objects(sim)
+        for obj in all_objects:
+            self.register_object(obj)
+
+    def register_object(
+        self, obj: Union[ManagedArticulatedObject, ManagedRigidObject]
+    ) -> None:
+        """
+        Register a single object in the 'objects_with_states' dict by checking 'is_affordance_of_obj' for all active ObjectStateSpecs.
+        Use this when a new object is added to the scene and needs to be registered.
+
+        :param obj: The ManagedObject instance to register.
+        """
+
+        for state in self.active_states:
+            if state.is_affordance_of_obj(obj):
+                if obj.handle not in self.objects_with_states:
```

Comment:
Not sure I understand this. Why not do:
`if obj.handle in self.objects_with_states:
      self.objects_with_states[obj.handle].append(state)`